psi4/driver/mdi_engine.py

```diff
@@ -0,0 +1,417 @@
+        return charges
+
+    # Respond to the <MASSES command
+    def send_masses(self):
+        """ Send the nuclear masses through MDI
+        """
+        natom = self.molecule.natom()
+        masses = [self.molecule.mass(iatom) for iatom in range(natom)]
+        MDI_Send(masses, natom, MDI_DOUBLE, self.comm)
+        return masses
+
+    # Respond to the <ELEMENTS command
+    def send_elements(self):
+        """ Send the atomic number of each nucleus through MDI
+        """
+        natom = self.molecule.natom()
+        elements = [self.molecule.true_atomic_number(iatom) for iatom in range(natom)]
+        MDI_Send(elements, natom, MDI_INT, self.comm)
+        return elements
+
+    # Respond to the <ENERGY command
+    def send_energy(self):
+        """ Send the total energy through MDI
+        """
+        MDI_Send(self.energy, 1, MDI_DOUBLE, self.comm)
+        return self.energy
+
+    # Respond to the <FORCES command
+    def send_forces(self):
+        """ Send the nuclear forces through MDI
+        """
+        force_matrix = psi4.driver.gradient(self.scf_method)
+        forces = force_matrix.np.ravel()
+        MDI_Send(forces, len(forces), MDI_DOUBLE_NUMPY, self.comm)
+        return forces
+
+    # Respond to the >CHARGES command
+    def recv_charges(self, charges=None):
+        """ Receive a set of nuclear charges through MDI and assign them to the atoms in the current molecule
+
+        Arguments:
+            charges: New nuclear charges. If None, receive through MDI.
+        """
+        natom = self.molecule.natom()
+        if charges is None:
+            charges = MDI_Recv(natom, MDI_DOUBLE, self.comm)
+        for iatom in range(natom):
+            self.molecule.set_nuclear_charge(iatom, charges[iatom])
+
+    # Respond to the >COORDS command
+    def recv_coords(self, coords=None):
+        """ Receive a set of nuclear coordinates through MDI and assign them to the atoms in the current molecule
+
+        Arguments:
+            coords: New nuclear coordinates. If None, receive through MDI.
+        """
+        natom = self.molecule.natom()
+        if coords is None:
+            coords = MDI_Recv(3 * natom, MDI_DOUBLE, self.comm)
+        matrix = psi4.core.Matrix.from_array(np.array(coords).reshape(-1, 3))
+        self.molecule.set_geometry(matrix)
+
+    # Respond to the >MASSES command
+    def recv_masses(self, masses=None):
+        """ Receive a set of nuclear masses through MDI and assign them to the atoms in the current molecule
+
+        Arguments:
+            masses: New nuclear masses. If None, receive through MDI.
+        """
+        natom = self.molecule.natom()
+        if masses is None:
+            masses = MDI_Recv(natom, MDI_DOUBLE, self.comm)
+
+        # Assign the mass of all atoms, taking care to avoid ghost atoms
+        jatom = 0
+        for iatom in range(natom):
+            while self.molecule.fcharge(jatom) == 0.0 and jatom < self.molecule.nallatom():
```

Comment:
strangely, Z=0 or atom in "ghost" fragment is the best test of ghostedness, rather than charge. Why work around ghosts here for mass but not in above fn for charges?